File: 백준/Gold/2294. 동전 2/동전 2.cc
#include <iostream>

using namespace std;

int main() {
    int n, k;
    
    cin >> n >> k;
    
    int coin[n];
    int dp[10001];
    
    for (int i = 0; i < n; i++) {
        cin >> coin[i];
    }
    
    for (int i = 1; i <= k; i++) {
        dp[i] = 10001;
    }
    
    for (int i = 0; i < n; i++) {
        for (int j = coin[i]; j <= k; j++) {
            dp[j] = min(dp[j], dp[j - coin[i]] + 1);
        }
    }
    
    if (dp[k] == 10001) cout << -1 << '\n';
    else cout << dp[k] << '\n';
    
    return 0;
}

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        
        int[] coin = new int[101];
        int[] dp = new int[10001];
        
        for (int i = 0; i < n; i++) {
            coin[i] = Integer.parseInt(br.readLine());
        }
        
        for (int i = 1; i <= k; i++) {
            dp[i] = 10001;
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = coin[i]; j <= k; j++) {
                dp[j] = Math.min(dp[j], dp[j - coin[i]] + 1);
            }
        }
        
        if (dp[k] == 10001) System.out.println(-1);
        else System.out.println(dp[k]);
    }
}